src/lib/pubkey/dilithium/dilithium.cpp

```diff
@@ -0,0 +1,749 @@
+            {
+               continue;
+            }
+
+            /* Compute hints for w1 */
+            t0.polyvec_pointwise_poly_montgomery(h, cp);
+            h.invntt_tomont();
+            h.reduce();
+            if (h.polyvec_chknorm(mode.get_gamma2()))
+            {
+               continue;
+            }
+
+            std::get<1>(w1_w0).add_polyvec(h);
+            std::get<1>(w1_w0).cadd_q();
+               
+            unsigned int n = PolynomialVector::generate_hint_polyvec(h, std::get<1>(w1_w0), std::get<0>(w1_w0), mode);
+            if (n > mode.get_omega())
+            {
+               continue;
+            }
+
+            /* Write signature */
+            packed_sig = pack_sig(sm.data(), z, h);
+            go_to_rej = false;
+         }
+         return packed_sig;
+      }
+
+      size_t signature_length() const override
+      {
+         auto& dilithium_math = m_priv_key.m_private->get_mode();
+         return dilithium_math.get_crypto_bytes();
+      }
+   private:
+
+      // Bit-pack signature sig = (c, z, h).
+      secure_vector<uint8_t> pack_sig(const uint8_t* c,
+         const PolynomialVector& z, const PolynomialVector& h)
+      {
+         size_t position = 0;
+         auto mode = m_priv_key.m_private->get_mode();
+         secure_vector<uint8_t> sig(mode.get_crypto_bytes());
+         for ( size_t i = 0; i < DilithiumModeConstants::SEEDBYTES; ++i)
+         {
+            sig[i] = c[i];
+         }
+         position += DilithiumModeConstants::SEEDBYTES;
+
+         for ( size_t i = 0; i < mode.get_l(); ++i)
+         {
+            z.m_vec[i].polyz_pack(&sig[position + i*mode.get_polyz_packedbytes()], mode);
+         }
+         position += mode.get_l()*mode.get_polyz_packedbytes();
+
+         /* Encode h */
+         for ( size_t i = 0; i < mode.get_omega() + mode.get_k(); ++i)
+         {
+            sig[i + position] = 0;
+         }
+
+         size_t k = 0;
+         for ( size_t i = 0; i < mode.get_k(); ++i)
+         {
+            for ( size_t j = 0; j < DilithiumModeConstants::N; ++j)
+               if (h.m_vec[i].m_coeffs[j] != 0)
+               {
+                  sig[position + k] = static_cast<uint8_t>(j);
+                  k++;
+               }
+            sig[position + mode.get_omega() + i] = static_cast<uint8_t>(k);
+         }
+         return sig; 
+      }
+
+      const Dilithium_PrivateKey& m_priv_key;
+      std::vector<uint8_t> m_msg;
+
+   };
+
+   class Dilithium_Verification_Operation final : public PK_Ops::Verification
+   {
+   public:
+      Dilithium_Verification_Operation(const Dilithium_PublicKey& pub_dilithium) : m_pub_key (pub_dilithium)
+      {
+      }
+      /*
+      * Add more data to the message currently being signed
+      * @param msg the message
+      * @param msg_len the length of msg in bytes
+      */
+      void update(const uint8_t msg[], size_t msg_len) override 
+      {
+         m_msg.insert(m_msg.end(), msg, msg + msg_len);
```

Comment:
Just like the signing operation: We should check whether we can take advantage of the streaming interface and avoid copying the entire message.